designs/disruption-termination-grace-period.md

```diff
@@ -0,0 +1,122 @@
+    terminationGracePeriod *NillableDuration `json:"terminationGracePeriod" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+The `terminationGracePeriod` fields accepts a common duration string which defaults to a value of `nil`. Omitting the field results in the default nil value instructing the controller to wait indefinitely for pods to drain gracefully, maintaining the existing behavior.
+A value of 0 instructs the controller to evict all pods by force immediately, matching the behavior of `pod.spec.terminationGracePeriod`
+
+## Prior Art
+This is the field from CAPI's MachineDeployment spec which implements similar behavior.
+```
+$ kubectl explain machinedeployment.spec.template.spec.nodeDrainTimeout
+KIND:     MachineDeployment
+VERSION:  cluster.x-k8s.io/v1beta1
+
+FIELD:    nodeDrainTimeout <string>
+
+DESCRIPTION:
+     NodeDrainTimeout is the total amount of time that the controller will spend
+     on draining a node. The default value is 0, meaning that the node can be
+     drained without any time limitations. NOTE: NodeDrainTimeout is different
+     from `kubectl drain --timeout`
+```
+
+```
+$ kubectl explain pod.spec.terminationGracePeriod
+KIND:     Pod
+VERSION:  v1
+
+FIELD:    terminationGracePeriod <integer>
+
+DESCRIPTION:
+     Optional duration in seconds the pod needs to terminate gracefully. May be
+     decreased in delete request. Value must be non-negative integer. The value
+     zero indicates stop immediately via the kill signal (no opportunity to shut
+     down). If this value is nil, the default grace period will be used instead.
+     The grace period is the duration in seconds after the processes running in
+     the pod are sent a termination signal and the time when the processes are
+     forcibly halted with a kill signal. Set this value longer than the expected
+     cleanup time for your process. Defaults to 30 seconds.
+```
+
+## Implementation
+
+### Termination Grace Period Expiration Detection
+Because Karpenter already ensures that nodes and their nodeclaims are in a deleting state before performing a node drain during node deletion, we should be able to leverage existing `deletionTimestamp` fields to avoid the need for an addition annotation or other tracking label.
+I believe we should use the NodeClaim `deletionTimestamp` specifically to avoid depending on the Node's `deletionTimestamp`, because [Kubernetes docs](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) recommend draining a node *before* deleting it, which is counter to how Karpenter behaves today (relying on Node finalizers to handle cleanup).
+
+If drains are not already periodically requeued, we may need to modify the current drain wait logic to periodically check if a node has exceeded its termination grace period.
+
+### Termination Grace Period Expiration Behavior
+1. Node deletion occurs (user initiated, node rotation from drift, etc).
```

Comment:
To clarify Karpenter's process here: Karpenter initiates a deletion on the NodeClaim first before it initiates the delete of the Node. This might be slightly different if a user initiates the delete on the node directly, but in the automated way, the ordering is:
1. Karpenter initiates NodeClaim deletion
2. Karpenter initiates Node deletion which triggers the drain

Reply:
The first thing the node deletion finalizer does is delete all NodeClaims for the Node that's being deleted, so I think both of those paths should operate identically for the purposes of this feature.
https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/node/termination/controller.go#L76-L82

Reply:
Updated the expected behavior here to cover both cases (NodeClaim vs Node deleted first).